chronicles/topics_registry.nim

```diff
@@ -12,84 +12,115 @@ type
-  for val in registry.topicStatesTable.values:
-    val.state = Normal
-
-iterator topicStates*: (string, TopicState) =
-  for name, topic in registry.topicStatesTable:
-    yield (name, topic.state)
-
-proc registerTopic*(name: string, topic: ptr Topic): ptr Topic =
-  registry.topicStatesTable[name] = topic
-  return topic
-
-proc setTopicState*(name: string,
-                    newState: TopicState,
-                    logLevel = LogLevel.NONE): bool =
-  if not registry.topicStatesTable.hasKey(name):
-    return false
-
-  var topicPtr = registry.topicStatesTable[name]
-
-  case topicPtr.state
-  of Enabled: dec registry.totalEnabledTopics
-  of Required: dec registry.totalRequiredTopics
-  else: discard
-
-  case newState
-  of Enabled: inc registry.totalEnabledTopics
-  of Required: inc registry.totalRequiredTopics
-  else: discard
-
-  topicPtr.state = newState
-  topicPtr.logLevel = logLevel
-
-  return true
-
-proc topicsMatch*(logStmtLevel: LogLevel,
-                  logStmtTopics: openarray[ptr Topic]): bool =
-  var
-    hasEnabledTopics = registry.totalEnabledTopics > 0
-    enabledTopicsMatch = false
-    normalTopicsMatch = logStmtTopics.len == 0
-    requiredTopicsCount = registry.totalRequiredTopics
-
-  for topic in logStmtTopics:
-    let topicLogLevel = if topic.logLevel != NONE: topic.logLevel
-                        else: gActiveLogLevel
-    if logStmtLevel >= topicLogLevel:
-      case topic.state
-      of Normal: normalTopicsMatch = true
-      of Enabled: enabledTopicsMatch = true
-      of Disabled: return false
-      of Required: dec requiredTopicsCount
+  lockRegistry:
+    registry.totalEnabledTopics = 0
+    registry.totalRequiredTopics = 0
+    registry.topicStatesTable.clear()
 
-  if requiredTopicsCount > 0:
-    return false
+proc registerTopic*(name: string, topic: Topic) =
+  lockRegistry:
+    registry.topicStatesTable[name] = topic
 
-  if hasEnabledTopics and not enabledTopicsMatch:
-    return false
+proc setTopicState*(name: string, newState: TopicState, logLevel = LogLevel.NONE): bool =
+  result = true
+  lockRegistry:
+    block registryBlock: # for early exits without bypassing the lock release
+      if not registry.topicStatesTable.hasKey(name):
+        result = false
+        break
 
-  return normalTopicsMatch
+      var topic = registry.topicStatesTable[name]
 
-proc getTopicState*(topic: string): ptr Topic =
-  return registry.topicStatesTable.getOrDefault(topic)
+      case topic.state
+      of Enabled: dec registry.totalEnabledTopics
+      of Required: dec registry.totalRequiredTopics
+      else: discard
+
+      case newState
+      of Enabled: inc registry.totalEnabledTopics
+      of Required: inc registry.totalRequiredTopics
+      else: discard
+
+      topic.state = newState
+      topic.logLevel = logLevel
+      registry.topicStatesTable[name] = topic
+
+proc topicsMatch*(logStmtLevel: LogLevel, logStmtTopics: openarray[Topic]): bool =
+  lockRegistry:
```

Comment:
This shouldn't lock either. All the information needed by the function is passed in as parameters.

Reply:
> All the information needed by the function is passed in as parameters.

It still reads from 2 global variables that may be changed at the same time from other threads.

Reply:
Hrm, ok, I'll think about this. It's OK to merge with locking now